components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/AbstractOTPAuthenticator.java

```diff
@@ -0,0 +1,1243 @@
+        }
+    }
+
+    /**
+     * Handle OTP resend attempts count exceeded scenarios.
+     *
+     * @param request  HttpServletRequest.
+     * @param response HttpServletResponse.
+     * @param context  AuthenticationContext.
+     * @throws AuthenticationFailedException If an error occurred.
+     */
+    private void handleOTPResendCountExceededScenario(HttpServletRequest request,
+                                                      HttpServletResponse response, AuthenticationContext context)
+            throws AuthenticationFailedException {
+
+        String queryParams = FrameworkUtils.getQueryStringWithFrameworkContextId(context.getQueryParams(),
+                context.getCallerSessionKey(), context.getContextIdentifier());
+        redirectToErrorPage(request, response, context, queryParams, ERROR_USER_RESEND_COUNT_EXCEEDED_QUERY_PARAMS);
+    }
+
+    protected AuthenticationScenarios resolveScenario(HttpServletRequest request,
+                                                                                        AuthenticationContext context) {
+
+        if (context.isLogoutRequest()) {
+            return LOGOUT;
+        } else if (!context.isRetrying() && StringUtils.isBlank(request.getParameter(CODE)) &&
+                StringUtils.isBlank(request.getParameter(RESEND))) {
+            return INITIAL_OTP;
+        } else if (context.isRetrying() && StringUtils.isNotBlank(request.getParameter(RESEND)) &&
+                Boolean.parseBoolean(request.getParameter(RESEND))) {
+            return RESEND_OTP;
+        }
+        return SUBMIT_OTP;
+    }
+
+    /**
+     * To redirect the flow to otp login page to enter an OTP.
+     *
+     * @param response      HttpServletResponse.
+     * @param request       HttpServletRequest.
+     * @param context       AuthenticationContext.
+     * @throws AuthenticationFailedException If an error occurred while redirecting to otp login page.
+     */
+    protected void redirectToOTPLoginPage(AuthenticatedUser authenticatedUser, String tenantDomain,
+                                          boolean isInitialFederationAttempt, HttpServletResponse response,
+                                          HttpServletRequest request, AuthenticationContext context)
+            throws AuthenticationFailedException {
+
+        String username = authenticatedUser.getUserName();
+        StringBuilder queryParamsBuilder = new StringBuilder();
+        String queryParams = FrameworkUtils.getQueryStringWithFrameworkContextId(context.getQueryParams(),
+                context.getCallerSessionKey(), context.getContextIdentifier());
+        String multiOptionURI = AbstractOTPAuthenticatorUtils.getMultiOptionURIQueryParam(request);
+        queryParamsBuilder.append(queryParams)
+                .append(AUTHENTICATORS_QUERY_PARAM).append(getName())
+                .append("&username=").append(username)
```

Comment:
Since you have already extracted constants for query params, do the same for this one as well.